[Prototype]Basic implementation of torch.distributed state_dict save/load

zzzwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71424

The API is mostly following
https://docs.google.com/document/d/1PGwGxVVVk6lsAWgtyKJnUORHg1wAR8XJFntFS1Nq9ZU/edit


This diff write/load state dict to file without any resharding.

Differential Revision: [D31908125](https://our.internmc.facebook.com/intern/diff/D31908125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31908125/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Commit:
[Prototype]Basic implementation of torch.distributed state_dict save/load

The API is mostly following
https://docs.google.com/document/d/1PGwGxVVVk6lsAWgtyKJnUORHg1wAR8XJFntFS1Nq9ZU/edit


This diff write/load state dict to file without any resharding.

Differential Revision: [D31908125](https://our.internmc.facebook.com/intern/diff/D31908125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31908125/)!

[ghstack-poisoned]

Commit:
Update on "[Prototype]Basic implementation of torch.distributed state_dict save/load"


The API is mostly following
https://docs.google.com/document/d/1PGwGxVVVk6lsAWgtyKJnUORHg1wAR8XJFntFS1Nq9ZU/edit


This diff write/load state dict to file without any resharding.

Differential Revision: [D31908125](https://our.internmc.facebook.com/intern/diff/D31908125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31908125/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[Prototype]Basic implementation of torch.distributed state_dict save/load"


The API is mostly following
https://docs.google.com/document/d/1PGwGxVVVk6lsAWgtyKJnUORHg1wAR8XJFntFS1Nq9ZU/edit


This diff write/load state dict to file without any resharding.

Differential Revision: [D31908125](https://our.internmc.facebook.com/intern/diff/D31908125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31908125/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

mrshenli commented on torch/distributed/_checkpoint/storage_reader.py:
```diff
@@ -0,0 +1,58 @@
+
+    def read(self, req: ReadWriteRequest) -> Future:
+        """
+        Very basic implementation that write to file system.
```
nit: read from?

sisilmehta2000 commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,45 @@
+
+@dataclass
+class StorageMetadata:
+    shard_metadata: Optional[ShardMetadata]
```
ShardMetadata is repeated in both StorageMetadata and TensorMetadata. Could it be isolated to the TensorMetadata only (since they are both contained in ExtendedTensorMetadata. It seems the ExtendedTensorMetadata is the overarching metadata container for saving/loading tensors whether they are distributed or not

zzzwen replied:
One is ShardMetadata the other is ShardedTensorMetadata, they are both define in "shard.sharded_tenosr". IMO the name can be confusing, might worth some update, @pritamdamania, @wanchaol

sisilmehta2000 commented on torch/distributed/_checkpoint/storage_reader.py:
```diff
@@ -0,0 +1,58 @@
+        pass
+
+class FileSystemReader(StorageReader):
+    def __init__(self, base_folder_name: str) -> None:
```
'path' instead of base_folder_name

sisilmehta2000 commented on torch/distributed/_checkpoint/storage_reader.py:
```diff
@@ -0,0 +1,58 @@
+        Very basic implementation that write to file system.
+        Note that the future is resolved before returning.
+        """
+        with open(f"{self.base_folder_name}/{req.storage_key}", "rb") as storage:
```
os.path.join()

sisilmehta2000 commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,72 @@
+            # They should be updated.
+            # Somehow I need to cast the memoryview to "c"ontiguous
+            # otherwise I will get a size mismatch at writing
+            mv = memoryview(req.target_tensor.detach().cpu().numpy()).cast("c")
```
here, how about we pickle the numpy object and get its size and write that to file? That way numpy metadata will be written as well. I'm wondering if here getting the memoryview of the numpy object gets the underlying array losing metadata info like dtype. (Which is prolly why you can't load the original dtype on reading back the saved data)

kumpera commented on test/distributed/_checkpoint/test_state_dict_saver_file_system.py:
```diff
@@ -0,0 +1,148 @@
+class TestStateDictSaveLoad(TestCase):
+    def test_read_write_only_tensor(self):
+        state_dict_to_save = TestModule().state_dict()
+        base_dir = "/tmp/_test_state_dict_save_load_torch_tensor_"
```
Maybe use https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory which ensures we won't clash with other processes.

kumpera commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,213 @@
+            # The assumption is that non ShardedTensors are full replicated across all ranks
+            # So we just need one from Rank 0.
+            # If that's not the case, we will update later.
+            if dist.is_initialized() and dist.get_rank() != 0:
```
Shouldn't we try to minimize the load across all ranks?
Something like `hash(fqn) % dist.world_size() != dist.get_world_size()` instead?

zzzwen replied:
Of course, I am just doing whatever that is simplest.  There will be some implication on the call sequence w.r.t. checkpoint load and  DDP, alltoall vs broadcast. This will require more discussion.

zzzwen commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,95 @@
+# compatible.
+#
+# The default resharding is defined in `reshard_and_prepare_fn`
+def load_state_dict(
```
IMO, the versioning comes from `torch.distributed`, what's the general guidance for versioning and BC for code in torch.distributed?

Commit:
Update on "[Prototype]Basic implementation of torch.distributed state_dict save/load"


The API is mostly following
https://docs.google.com/document/d/1PGwGxVVVk6lsAWgtyKJnUORHg1wAR8XJFntFS1Nq9ZU/edit


This diff write/load state dict to file without any resharding.

Differential Revision: [D31908125](https://our.internmc.facebook.com/intern/diff/D31908125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31908125/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[Prototype]Basic implementation of torch.distributed state_dict save/load"


The API is mostly following
https://docs.google.com/document/d/1PGwGxVVVk6lsAWgtyKJnUORHg1wAR8XJFntFS1Nq9ZU/edit


This diff write/load state dict to file without any resharding.

Differential Revision: [D31908125](https://our.internmc.facebook.com/intern/diff/D31908125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31908125/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

wilson100hong commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,47 @@
+        self.state_dict_metadata = pickle.loads(state)
+
+@dataclass
+class ReadWriteRequest:
```
can we separate to ReadRequest and WriteRequest to make it simpler?
For example, do we need `length` for writing request? Which I assume the length can be drivied fro `target_tensor`

wilson100hong commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,72 @@
+
+class StorageWriter(abc.ABC):
+    """
+    Interface to write to underlying storage system
```
nit: period at the end '.'

wilson100hong commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,72 @@
+    Interface to write to underlying storage system
+    """
+    @abc.abstractmethod
+    def write(self, req: ReadWriteRequest) -> Future:
```
would it be convenient to provide sync write() ?

wilson100hong commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,72 @@
+        pass
+
+    @abc.abstractmethod
+    def write_metadata(self, metadata: Metadata) -> None:
```
qq, why write() is async but write_metadata is sync()? Are they written to different destinations?

wilson100hong commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,214 @@
+def save_state_dict_async(
+    state_dict: Dict[str, Any],
+    storage_writer: StorageWriter,
+    metadata_prepare_fn: Callable = _prepare,
```
Should we add type annotation for like `Callable[[Dict[str,Any]], Tuple(Any,Any,Any)]` ?

Commit:
Update on "[Prototype]Basic implementation of torch.distributed state_dict save/load"


The API is mostly following
https://docs.google.com/document/d/1PGwGxVVVk6lsAWgtyKJnUORHg1wAR8XJFntFS1Nq9ZU/edit


This diff write/load state dict to file without any resharding.

Differential Revision: [D31908125](https://our.internmc.facebook.com/intern/diff/D31908125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31908125/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,47 @@
+        self.state_dict_metadata = pickle.loads(state)
+
+@dataclass
+class ReadWriteRequest:
```
This is not actually `metadata`? Probably belongs in the StorageReader/Writer classes. Also, this will probably have to move to C++ given that we want to replace `torch.Tensor` with a buffer.

pritamdamania87 commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,47 @@
+    # Metadata for the state dict.
+    state_dict_metadata: Dict[str, ExtendedTensorMetadata]
+
+    def __getstate__(self) -> bytes:
+        serialized = pickle.dumps(self.state_dict_metadata)
+        return serialized
+
+    def __setstate__(self, state: bytes):
+        # TODO, use pickle for this quick hack, must replace it with something else e.g. flatbuffer
+        # before serious use case,
+        self.state_dict_metadata = pickle.loads(state)
```
Do we really need to do this? Shouldn't dataclass automatically generate `__getstate__` and `__setstate__` to enable pickling?

pritamdamania87 commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,47 @@
+@dataclass
+class ExtendedTensorMetadata:
+    # Details of Tensor/ShardedTensor (dtype, shape, sharding config etc.)
+    tensor_metadata: ShardedTensorMetadata
```
If we have a regular `torch.Tensor`, how do we represent its metadata here?

zzzwen replied:
I would think there need be an equivalent to ShardedTensorMetadata, let me add some inline comment

pritamdamania87 commented on torch/distributed/_checkpoint/metadata.py:
```diff
@@ -0,0 +1,47 @@
+TENSOR_TYPE = Union[torch.Tensor, ShardedTensor]
+
+@dataclass
+class StorageMetadata:
```
All of this will probably change when we actually use flatbuffers for the metadata?

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,96 @@
+    torch.futures.wait_all(futures)
+
+
+def load_state_dict_async(
```
Why do we need an async variant of this? Let us try to keep the API minimal and only have sync API for now.

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,96 @@
+        storage_reader (StorageReader): An instance of storage loader.
+        reshard_and_prepare_fn (Callable): The resharding function.
+    """
+    metadata = storage_reader.read_metadata()
```
storage_reader shouldn't worry about metadata, we probably should use a special storage_reader key, and have something like `storage_reader.read('.metadata')`.

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,96 @@
+    read_requests = []
+    for fqn, tensor in state_dict.items():
+        if isinstance(tensor, torch.Tensor):
+            # Reshape to 1D, the might make loading eaiser
+            target_tensor = tensor.detach().reshape(-1)
+            storage_size = target_tensor.nelement() * target_tensor.element_size()
```
Why do we need to reshape like this? The underlying storage of the Tensor is anyways a contiguous buffer that we can use.

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,96 @@
+            read_requests.append(rr)
+        elif isinstance(tensor, ShardedTensor):
+            for shard in tensor.local_shards():
+                suffix = "_".join([str(i) for i in shard.metadata.shard_offsets + shard.metadata.shard_sizes])
```
Isn't `shard.metadata.shard_offsets` enough to generate a unique key? Why do we need to use `shard_sizes` as well?

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,213 @@
+
+    torch.futures.wait_all(futures)
+
+def save_state_dict_async(
```
Same as above, let us just have a sync version to begin with.

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_loader.py:
```diff
@@ -0,0 +1,96 @@
+def load_state_dict(
+    state_dict: Dict[str, Any],
+    storage_reader: StorageReader,
+    reshard_and_prepare_fn: Callable = _reshard_and_prepare_read_request,
```
What is the reason that this is a Callable that the user can configure instead of just calling `_reshard_and_prepare_read_request` inline?

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,213 @@
+def save_state_dict(
+    state_dict: Dict[str, Any],
+    storage_writer: StorageWriter,
+    metadata_prepare_fn : Callable = _prepare,
```
What is the need for this to be configurable by the user?

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,213 @@
+    )
+
+    etmd = ExtendedTensorMetadata(
+        tensor_metadata=stm,
```
For the torch.Tensor case, can't tensor_metadata be `Union[ShardedTensorMetadata, TensorProperties]`? Or probably we need to create another class like `TensorMetadata` which includes size as well since TensorProperties doesn't have size.

Reusing ShardedTensorMetadata here doesn't seem appropriate.

pritamdamania87 commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,72 @@
+        """
+        pass
+
+    def prepare_storage(self, storage_handles: Dict[str, int]) -> None:
```
What is this API needed for? We can always add it later if needed?

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,213 @@
+):
+    smd = []
+    for shard in tensor.local_shards():
+        # each shard is in it own file.
```
nit: In its own `storage_key`, the storage system can use however many files it wants under the hood.

pritamdamania87 commented on torch/distributed/_checkpoint/state_dict_saver.py:
```diff
@@ -0,0 +1,213 @@
+        # Most network file system is optimized with single write, multiple read
+        # Unless we can group tensors locally into one big chunk
+        # It might be best to write each shard as one key
+        suffix = "_".join([str(i) for i in shard.metadata.shard_offsets + shard.metadata.shard_sizes])
```
Don't need sizes here?

pritamdamania87 commented on torch/distributed/_checkpoint/storage_reader.py:
```diff
@@ -0,0 +1,59 @@
+import pickle
+import numpy
+
+class StorageReader(abc.ABC):
```
These interfaces need to move to C++?

pritamdamania87 commented on torch/distributed/_checkpoint/storage_reader.py:
```diff
@@ -0,0 +1,59 @@
+        pass
+
+    @abc.abstractmethod
+    def read_metadata(self) -> Metadata:
```
As mentioned above, we don't need this interface.

pritamdamania87 commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,72 @@
+import torch.distributed as dist
+import abc
+
+class StorageWriter(abc.ABC):
```
Same as above, these interfaces need to move to C++

pritamdamania87 commented on torch/distributed/_checkpoint/storage_writer.py:
```diff
@@ -0,0 +1,72 @@
+        pass
+
+    @abc.abstractmethod
+    def write_metadata(self, metadata: Metadata) -> None:
```
We shouldn't need `write_metadata`, just do `write('.metadata')` to a special metadata storage_key.
